Support native compilation on big-endian ppc64

When I try to compile the OCaml compiler from Git trunk on my POWER9 system the configuration step fails:

```console
$ ./configure --enable-native-compiler
...
configure: error: the native compiler is not supported on this platform
```

Checking the configure script suggests that only `powerpc64le*-*-linux*` is whitelisted with `has_native_backend=yes`.

Since the README says that native compilation on the Power architecture will be supported again in OCaml 5.2, please also add `powerpc64-*-linux-*` to the list of supported native backends.

--- a/Changes
+++ b/Changes
@@ -1,11 +1,14 @@
 Working version
 ---------------
 
-### Restored backends:
+### Restored and new backends:
 
 - #12276, #12601: native-code compilation for POWER (64 bits, little-endian)
   (Xavier Leroy, review by KC Sivaramakrishnan, Anil Madhavapeddy,
    and Stephen Dolan)
+
+- #12667: extend the latter to POWER 64 bits, big-endian, ELFv2 ABI
+  (A. Wilcox, review by Xavier Leroy)
 
 ### Language features:
 

--- a/README.adoc
+++ b/README.adoc
@@ -71,7 +71,7 @@ compiler currently runs on the following platforms:
 
 | x86 64 bits    | Linux, macOS, Windows, FreeBSD  |  NetBSD, OpenBSD, OmniOS (Solaris)
 | ARM 64 bits    | Linux, macOS                    |  FreeBSD, OpenBSD, NetBSD
-| Power 64 bits  | Linux                           |
+| Power 64 bits  | Linux (little-endian, ABIv2)    |  Linux (big-endian, ABIv2)
 | RISC-V 64 bits | Linux                           |
 | IBM Z (s390x)  | Linux                           |
 |====

--- a/asmcomp/power/arch.ml
+++ b/asmcomp/power/arch.ml
@@ -47,7 +47,11 @@ type addressing_mode =
 
 (* Sizes, endianness *)
 
-let big_endian = false (* ppc64le only *)
+let big_endian =
+  match Config.model with
+  | "ppc64" -> true
+  | "ppc64le" -> false
+  | _ -> assert false
 let size_addr = 8
 let size_int = size_addr
 let size_float = 8

--- a/configure
+++ b/configure
@@ -15493,6 +15493,8 @@ case $host in #(
     arch=amd64; system=win64 ;; #(
   powerpc64le*-*-linux*) :
     has_native_backend=yes; arch=power; model=ppc64le; system=linux ;; #(
+  powerpc64*-*-linux-musl*) :
+    has_native_backend=yes; arch=power; model=ppc64; system=linux ;; #(
   s390x*-*-linux*) :
     has_native_backend=yes; arch=s390x; model=z10; system=linux ;; #(
   # expected to match "gnueabihf" as well as "musleabihf"

--- a/configure.ac
+++ b/configure.ac
@@ -1262,6 +1262,8 @@ AS_CASE([$host],
     [arch=amd64; system=win64],
   [[powerpc64le*-*-linux*]],
     [has_native_backend=yes; arch=power; model=ppc64le; system=linux],
+  [[powerpc64*-*-linux-musl*]],
+    [has_native_backend=yes; arch=power; model=ppc64; system=linux],
   [[s390x*-*-linux*]],
     [has_native_backend=yes; arch=s390x; model=z10; system=linux],
   # expected to match "gnueabihf" as well as "musleabihf"

--- a/runtime/power.S
+++ b/runtime/power.S
@@ -15,6 +15,10 @@
 
         .abiversion 2
 
+#if _CALL_ELF != 2
+#error "This POWER port requires the ELFv2 ABI"
+#endif
+
 /* Special registers */
 #define SP 1
 #define TMP 11

--- a/testsuite/tools/asmgen_power.S
+++ b/testsuite/tools/asmgen_power.S
@@ -10,55 +10,11 @@
 /*                                                                   */
 /*********************************************************************/
 
-#if defined(MODEL_ppc64) || defined(MODEL_ppc64le)
-#define EITHER(a,b) b
-#else
-#define EITHER(a,b) a
-#endif
-
-#define WORD EITHER(4,8)
-#define lg EITHER(lwz,ld)
-#define lgu EITHER(lwzu,ldu)
-#define stg EITHER(stw,std)
-#define stgu EITHER(stwu,stdu)
-
-#if defined(MODEL_ppc)
-#define RESERVED_STACK 16
-#define LR_SAVE_AREA 4
-#endif
-#if defined(MODEL_ppc64)
-#define RESERVED_STACK 48
-#define LR_SAVE_AREA 16
-#endif
-#if defined(MODEL_ppc64le)
 #define RESERVED_STACK 32
 #define LR_SAVE_AREA 16
-#endif
 
 /* Function definitions */
 
-#if defined(MODEL_ppc)
-#define FUNCTION(name) \
-  .section ".text"; \
-  .globl name; \
-  .type name, @function; \
-  .align 2; \
-  name:
-#endif
-
-#if defined(MODEL_ppc64)
-#define FUNCTION(name) \
-  .section ".opd","aw"; \
-  .align 3; \
-  .globl name; \
-  .type name, @function; \
-  name: .quad .L.name,.TOC.@tocbase; \
-  .text; \
-  .align 2; \
-  .L.name:
-#endif
-
-#if defined(MODEL_ppc64le)
 #define FUNCTION(name) \
   .section ".text"; \
   .globl name; \
@@ -68,35 +24,34 @@
   0: addis 2, 12, (.TOC. - 0b)@ha; \
   addi 2, 2, (.TOC. - 0b)@l; \
   .localentry name, . - 0b
-#endif
 
 FUNCTION(call_gen_code)
     /* Allocate and link stack frame */
-        stgu    1, -(WORD*18 + 8*18 + RESERVED_STACK)(1)
+        stdu    1, -(8*18 + 8*18 + RESERVED_STACK)(1)
     /* 18 saved GPRs, 18 saved FPRs */
     /* Save return address */
         mflr    0
-        stg     0, (WORD*18 + 8*18 + RESERVED_STACK + LR_SAVE_AREA)(1)
+        std     0, (8*18 + 8*18 + RESERVED_STACK + LR_SAVE_AREA)(1)
     /* Save all callee-save registers, starting at RESERVED_STACK */
-        addi    11, 1, RESERVED_STACK - WORD
-        stgu    14, WORD(11)
-        stgu    15, WORD(11)
-        stgu    16, WORD(11)
-        stgu    17, WORD(11)
-        stgu    18, WORD(11)
-        stgu    19, WORD(11)
-        stgu    20, WORD(11)
-        stgu    21, WORD(11)
-        stgu    22, WORD(11)
-        stgu    23, WORD(11)
-        stgu    24, WORD(11)
-        stgu    25, WORD(11)
-        stgu    26, WORD(11)
-        stgu    27, WORD(11)
-        stgu    28, WORD(11)
-        stgu    29, WORD(11)
-        stgu    30, WORD(11)
-        stgu    31, WORD(11)
+        addi    11, 1, RESERVED_STACK - 8
+        stdu    14, 8(11)
+        stdu    15, 8(11)
+        stdu    16, 8(11)
+        stdu    17, 8(11)
+        stdu    18, 8(11)
+        stdu    19, 8(11)
+        stdu    20, 8(11)
+        stdu    21, 8(11)
+        stdu    22, 8(11)
+        stdu    23, 8(11)
+        stdu    24, 8(11)
+        stdu    25, 8(11)
+        stdu    26, 8(11)
+        stdu    27, 8(11)
+        stdu    28, 8(11)
+        stdu    29, 8(11)
+        stdu    30, 8(11)
+        stdu    31, 8(11)
         stfdu   14, 8(11)
         stfdu   15, 8(11)
         stfdu   16, 8(11)
@@ -116,18 +71,8 @@ FUNCTION(call_gen_code)
         stfdu   30, 8(11)
         stfdu   31, 8(11)
     /* Get function pointer in CTR */
-#if defined(MODEL_ppc)
-        mtctr   3
-#elif defined(MODEL_ppc64)
-        ld      0, 0(3)
-        mtctr   0
-        ld      2, 8(3)
-#elif defined(MODEL_ppc64le)
         mtctr   3
         mr      12, 3
-#else
-#error "wrong MODEL"
-#endif
     /* Shuffle arguments */
         mr      3, 4
         mr      4, 5
@@ -136,25 +81,25 @@ FUNCTION(call_gen_code)
     /* Call the function */
         bctrl
     /* Restore callee-save registers */
-        addi    11, 1, RESERVED_STACK - WORD
-        lgu     14, WORD(11)
-        lgu     15, WORD(11)
-        lgu     16, WORD(11)
-        lgu     17, WORD(11)
-        lgu     18, WORD(11)
-        lgu     19, WORD(11)
-        lgu     20, WORD(11)
-        lgu     21, WORD(11)
-        lgu     22, WORD(11)
-        lgu     23, WORD(11)
-        lgu     24, WORD(11)
-        lgu     25, WORD(11)
-        lgu     26, WORD(11)
-        lgu     27, WORD(11)
-        lgu     28, WORD(11)
-        lgu     29, WORD(11)
-        lgu     30, WORD(11)
-        lgu     31, WORD(11)
+        addi    11, 1, RESERVED_STACK - 8
+        ldu     14, 8(11)
+        ldu     15, 8(11)
+        ldu     16, 8(11)
+        ldu     17, 8(11)
+        ldu     18, 8(11)
+        ldu     19, 8(11)
+        ldu     20, 8(11)
+        ldu     21, 8(11)
+        ldu     22, 8(11)
+        ldu     23, 8(11)
+        ldu     24, 8(11)
+        ldu     25, 8(11)
+        ldu     26, 8(11)
+        ldu     27, 8(11)
+        ldu     28, 8(11)
+        ldu     29, 8(11)
+        ldu     30, 8(11)
+        ldu     31, 8(11)
         lfdu    14, 8(11)
         lfdu    15, 8(11)
         lfdu    16, 8(11)
@@ -174,26 +119,16 @@ FUNCTION(call_gen_code)
         lfdu    30, 8(11)
         lfdu    31, 8(11)
     /* Reload return address */
-        lg      0, (WORD*18 + 8*18 + RESERVED_STACK + LR_SAVE_AREA)(1)
+        ld      0, (8*18 + 8*18 + RESERVED_STACK + LR_SAVE_AREA)(1)
         mtlr    0
     /* Return */
-        addi    1, 1, (WORD*18 + 8*18 + RESERVED_STACK)
+        addi    1, 1, (8*18 + 8*18 + RESERVED_STACK)
         blr
 
 FUNCTION(caml_c_call)
     /* Jump to C function (address in r28) */
-#if defined(MODEL_ppc)
-        mtctr   28
-#elif defined(MODEL_ppc64)
-        ld      0, 0(28)
-        mtctr   0
-        ld      2, 8(28)
-#elif defined(MODEL_ppc64le)
         mtctr   28
         mr      12, 28
-#else
-#error "wrong MODEL"
-#endif
         bctr
 
 /* Mark stack as non-executable */

